src/functions-and-arrays.js

```diff
@@ -103,17 +203,92 @@ const matrix = [
   [4, 42, 16, 73, 38, 25, 39, 11, 24, 94, 72, 18, 8, 46, 29, 32, 40, 62, 76, 36],
   [20, 69, 36, 41, 72, 30, 23, 88, 34, 62, 99, 69, 82, 67, 59, 85, 74, 4, 36, 16],
   [20, 73, 35, 29, 78, 31, 90, 1, 74, 31, 49, 71, 48, 86, 81, 16, 23, 57, 5, 54],
-  [1, 70, 54, 71, 83, 51, 54, 69, 16, 92, 33, 48, 61, 43, 52, 1, 89, 19, 67, 48]
+  [1, 70, 54, 71, 83, 51, 54, 69, 16, 92, 33, 48, 61, 43, 52, 1, 89, 19, 67, 48],
 ];
 
-function greatestProduct() {}
+// This is working X/Y
+function greatestProduct(input) {
+  let product = 0;
+  let resultX, resultY;
+  for (let i = 0; i < input.length; i++) {
+    for (let j = 0; j < input.length - 3; j++) {
+      resultX = multiplyingX(input, j, i);
+      resultY = multiplyingY(input, i, j);
+      if (resultX > product) {
+        product = resultX;
+      } else if (resultY > product) {
+        product = resultY;
+      }
+    }
+  }
+  return product;
+}
+
+// This is working D1/D2
+function greatestProductOfDiagonals(input) {
+  let product = 0;
+  let resultD1, resultD2;
+  for (let i = 0; i < input.length - 3; i++) {
+    for (let j = 0; j < input.length - 3; j++) {
+      resultD1 = multiplyingD1(input, j, i);
+      resultD2 = multiplyingD2(input, j, i);
+      if (resultD1 > product) {
+        product = resultD1;
+      } else if (resultD2 > product) {
+        product = resultD2;
+      }
+    }
+  }
+  return product;
+}
+
+// Both unified
+function allAxisGreatestProduct(input) {
+  return greatestProduct(input) > greatestProductOfDiagonals(input)
+    ? greatestProduct(input)
+    : greatestProductOfDiagonals(input);
+}
+
+console.log(allAxisGreatestProduct(matrix));
+
+function multiplyingY(arr, a, b) {
+  return arr[a][b] * arr[a][b + 1] * arr[a][b + 2] * arr[a][b + 3];
+}
+function multiplyingX(arr, a, b) {
+  return arr[a][b] * arr[a + 1][b] * arr[a + 2][b] * arr[a + 3][b];
+}
+
+function multiplyingD1(arr, a, b) {
+  return arr[a][b] * arr[a + 1][b + 1] * arr[a + 2][b + 2] * arr[a + 3][b + 3];
+}
 
+function multiplyingD2(arr, a, b) {
+  return arr[a][b + 3] * arr[a + 1][b + 2] * arr[a + 2][b + 1] * arr[a + 3][b];
+}
 
+// Extra Bonus - Matrix Generator
+function makeMatrix(width, height, lowestNumber, highestNumber) {
+  let createdMatrix = [];
+  for (let i = 0; i < width; i++) {
+    createdMatrix[i] = [];
+    for (let j = 0; j < height; j++) {
+      createdMatrix[i][j] = getRandomNumber(lowestNumber, highestNumber);
+    }
+  }
+  return createdMatrix;
+}
+
+function getRandomNumber(min, max) {
+  min = Math.ceil(min);
+  max = Math.floor(max);
+  return Math.floor(Math.random() * (max - min)) + min;
+}
 
+console.log(makeMatrix(10, 10, 1, 100));
```

Comment:
all of the above bonus iterations is exceptional work! :partying_face: really well done. you deserve a tap on the shoulder or a trophy or something :)

Reply:
Thank you very much for all your detailed comments! And the ["JS Black Magic"] :)